### DISCUSSION TITLE
Bypass accessor when using increment() or decrement() methods

### DISCUSSION THREAD

**Discussion by grimmdude** [laravel/framework#52694] [Ideas] [answered]
Hi Everyone,

I noticed that incrementing/decrementing a database column with `increment()`/`decrement()` causes any custom accessor on that column attribute to be executed.  In my case this is undesirable, so I ended up using an `update()` query instead.

It seems to me that accessors should never be executed when incrementing/decrementing a column, but maybe I'm missing a use case?  Curious your thoughts.

I believe this is where the accessor is being fired:
https://github.com/laravel/framework/blob/b7c628776754a7d861dd48ff18768e1c1f2d5f5f/src/Illuminate/Database/Eloquent/Model.php#L977-L979

PR submitted: https://github.com/laravel/framework/pull/52701

**Comment by ghost** [comment]
$this->{$column} is generating that. as it is a 'get/read'. 
It could be replaced by 
$this->getAttributes()[$column] ?? null

**Comment by rodrigopedra** [comment]
> It seems to me that accessors should never be executed when incrementing/decrementing a column

Why not? Can you elaborate?

If someone has an accessor to a JSON field stored in DB, then reading from the accessor would be mandatory.

**Reply by grimmdude** [reply]
Maybe I'm misunderstanding how `increment()` works then because I would expect it to operate purely as a database query, not based on the value of the instantiated model.

**Reply by rodrigopedra** [reply]
I guess it used to be like that some time ago. But if the increased usage of `Attribute` and casts, it was changed to take those into account.

In some sense, if it was a pure database operation, the assignment would have to be changed, as it is now it would trigger any mutator on the field. 

Thus, as it would trigger a field's mutator, reading the value from its accessor makes more sense.

**Reply by rodrigopedra** [reply, marked as answer]
If you want to use a pure database approach, you can use the underlying `Illuminate\Database\Query\Builder@increment()` method instead:

```php
User::query()->toBase()->increment('login_count');
```

https://github.com/laravel/framework/blob/a2c66ad997d583560836ae1c578bbbba6d52e26b/src/Illuminate/Database/Query/Builder.php#L3924-L3931

**Reply by grimmdude** [reply]
Got it, thank you for that explanation.  Basing off of model values has the potential for a race condition unless I'm mistaken.  I will definitely start using that underlying `increment()` method instead for the pure database approach.

**Reply by Sophist-UK** [reply]
If there is a potential for a race condition then this should be tested and if necessary fixed.

@grimmdude Perhaps you could explain (in simple terms that dunces like me can understand) why you think that this might create a race condition?

**Reply by grimmdude** [reply]
I feel like I still must be misunderstanding, but if it's just incrementing based on model values then here's a very contrived race condition example:

```php
// SELECT sent_emails FROM users where id = 1
// > 1

$u1 = User::find(1);
$u2 = User::find(1);

$u1->sent_emails = $u1->sent_emails + 1;
$u1->save();
// SELECT sent_emails FROM users where id = 1
// > 2

$u2->sent_emails = $u2->sent_emails + 1;
$u2->save();
// SELECT sent_emails FROM users where id = 1
// > 2 (should be 3)
```
You probably would never do this in a script, but it's possible to happen when processing a high volume of async jobs for instance.   

I'm confused though because this is not how I understood `increment()` to work, and everywhere I read seems to imply this is _not_ how it works.  But if it's relying on model accessors to update the value then it certainly seems like this is the method it's using.

**Reply by Sophist-UK** [reply]
Yes - this is a race condition, however this is always an issue in Laravel - for example I have two users running the following at the same time:
```
$u = User::find(1);
$u->sent_emails = $u1->sent_emails + 1;
$u->save();
```
You normally have to code this inside a transaction to avoid this.

**Reply by rodrigopedra** [reply]
The `Illuminate\Database\Query\Builder@increment()` avoids this as it builds an `UPDATE` query using the column itself.

But as @Sophist-UK noted, this is a limitation of the Active Record design pattern Laravel uses. As each instance is its own copy from the database record, and is also responsible for interacting with the database, each will act with their original attribute set.

By the way, the solution you proposed would still have the same issue, as it isn't re-reading from the database and operates on the copy each instance holds from the corresponding database record.

**Reply by grimmdude** [reply]
Got it, thanks I understand that now.  The difference between `Illuminate\Database\Eloquent\Model\increment()` and `Illuminate\Database\Query\Builder@increment()` was the source of my confusion.

This seems to extend to the community, as many search results I see describe `increment()` as using a `SET foo = foo + 1`  query opposed to basing on model values.  Maybe this merits some clearer documentation between the two?

**Reply by rodrigopedra** [reply]
You could give it a try, I think it would be good information to have in the docs

**Reply by ghost** [reply]
> If you want to use a pure database approach, you can use the underlying `Illuminate\Database\Query\Builder@increment()` method instead:
> 
> ```
> User::query()->toBase()->increment('login_count');
> ```
> 
> https://github.com/laravel/framework/blob/a2c66ad997d583560836ae1c578bbbba6d52e26b/src/Illuminate/Database/Query/Builder.php#L3924-L3931

Hi. I'm failing to understand how this is identifying the row that needs incrementing?
 User::query()->toBase()->increment('login_count');

**Reply by rodrigopedra** [reply]
Internally this method calls the `incrementEach` method. Take a look into its implementation. 

I'm on mobile now, so it is hard to link to it.

But basically it gets an array with the columns as keys and amount as values and iterates through them building the update expression

**Reply by ghost** [reply]
@rodrigopedra but the idea was to increment a column from one row from db not all, no?

**Reply by rodrigopedra** [reply]
Yes, my example was just to tell about the `Builder@increment()` method. Ideally, one would need to pair it with a `where` clause.

For example, the following expression...

```php
User::query()
    ->whereKey(Auth::id())
    ->toBase()
    ->increment('login_count');
```

... will generate this query:

```sql
UPDATE users
SET login_count = login_count + 1
WHERE id = ?;
```

The `?` binding will be replaced by the current logged-in user.

**Reply by ghost** [reply]
Ok. So please also mention that this will not trigger the updated event as it is a raw update. Thank you for the explanation. If events are needed my suggestion is doable in a macro.

**Reply by Sophist-UK** [reply]
Actually, on reflection I wonder whether this whole question is based on a possible misunderstanding of the role of Models.

In my head it seems that data starts in the database and you can then base your application on one of two approaches:
1. Doing all operations on the database using SQL queries and Query builder - what you might call a traditional approach. For every database update you either explicitly update the data you hold in memory or you reload the data;
2. Using Eloquent and models - where you copy the data into memory, process it there and then write it back to the database.

What you don't generally do is mix these approaches. So you wouldn't read information into a model, increment a field in the database and then write the model back to the database again.

**Reply by ghost** [reply]
@Sophist-UK I forgot to mention about the model reload after the raw update.

**Reply by grimmdude** [reply]
>Actually, on reflection I wonder whether this whole question is based on a possible misunderstanding of the role of Models.

@Sophist-UK I think I understand the role of models as you describe them, however I suppose I expected `Illuminate\Database\Eloquent\Model\increment()` to have some higher level benefit over just updating a model property and calling `save()`.

Only `Illuminate\Database\Query\Builder@increment()` is mentioned in the docs (under "Update Statements") so I think this could be contributing to the confusion: https://laravel.com/docs/11.x/queries#increment-and-decrement

Including a section in the https://laravel.com/docs/11.x/eloquent docs explaining the difference might help clarify the behavioral difference between the two.

**Reply by ghost** [reply]
@grimmdude wait, what? the increment should generate sql like column = column + 1  for atomic behavior....

**Reply by ghost** [reply]
Now I see that it might generate out of sync data between the model and the db . see https://github.com/laravel/framework/discussions/52732
